### PR TITLE
[Core] Increase the dashboard agent gRPC payload to 20Mb & batch the metrics to multiple gRPC requests

### LINKED ISSUE #40033: [Core] Metrics being lost for actors/tasks
### What happened + What you expected to happen

We have a Serve application deployed and periodically we see metrics vanish for some deployments after a while (see screenshot)

Screenshot 2023-10-02 at 5 52 19 PM

### Versions / Dependencies

Ray 2.7

### Reproduction script

N/A

### Issue Severity

High: It blocks me from completing my task.

### PR DESCRIPTION
**Pull request by rkooo567**
## Why are these changes needed?

It seems like if there are lots of metrics to be reported, the worker fails to report metrics because the default grpc server payload size is 4MB.

This PR fixes the issue using 2 ways

- Increase the default gprc server payload limit to 20MB. The Ray's default is 512MB, so I think increasing this a little bit is not risky.
- When there are lots of metrics, the metrics are reported in multiple gRPC requests. It seems like each metric is about 500~1KB bytes, so we set 1000 as the default batch size (< 1MB). Note that if the default is too small, it can cause issues because agent is not that great at handling lots of GRPC requests.

## Related issue number

Closes https://github.com/ray-project/ray/issues/40033

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Review by rkooo567 — COMMENTED** [review]
Note: I manually verified the env var works.

**Comment by iycheng** [comment]
I think maybe streaming important for transferring big chunks of data.

**Comment by rkooo567** [comment]
@iycheng we will also send multiple batch requests actually. I am figuring out what's the best default

File: dashboard/consts.py [hunk 1]
```diff
@@ -79,3 +79,7 @@
     "dashboard",
     "gcs",
 }
+# 20MB
```
Comment (rickyyx) [inline dashboard/consts.py:82]: nit: i wonder if we should put all the grpc related consts in a single file. I know there are some other configs that we override? like the keep alive timeouts.

File: python/ray/tests/test_task_metrics.py [hunk 2]
```diff
@@ -598,6 +599,63 @@ def g():
         )
 
 
+@pytest.mark.skipif(sys.platform == "win32", reason="Flaky on Windows. Timing out.")
+def test_metrics_batch(shutdown_only):
```
Comment (rickyyx) [inline python/ray/tests/test_task_metrics.py:603]: does this test fail before the patch?

File: src/ray/common/ray_config_def.h [hunk 3]
```diff
@@ -479,7 +479,8 @@ RAY_CONFIG(int64_t, ray_syncer_polling_buffer, 5)
 RAY_CONFIG(uint64_t, gcs_service_address_check_interval_milliseconds, 1000)
 
 /// The batch size for metrics export.
-RAY_CONFIG(int64_t, metrics_report_batch_size, 100)
+/// Normally each metrics is about < 1KB. 1000 means it is around 1MB.
```
Comment (rickyyx) [inline src/ray/common/ray_config_def.h:482]: what does the original batch size field control if there's already such a config?

**Review by rickyyx — COMMENTED** [review]
Why do we need both changes? 

Would changing the payload size limit not be efficient?

File: python/ray/tests/test_task_metrics.py (hunk 2 above)
Comment (rkooo567) [reply python/ray/tests/test_task_metrics.py:603]: The point is when you batch, it ensures the corretness. So it doesn't fail before the patch. I checked the log that it sends 1 metrics in multiple grpc requests. If we want to test this part, we should write the whole unit test suite.. (I can do it if you think it is a blocker, but it is a lot of work)

File: src/ray/common/ray_config_def.h (hunk 3 above)
Comment (rkooo567) [reply src/ray/common/ray_config_def.h:482]: that one is not used anywhere (maybe except Ant). And based on my initial research, 100 is anyway too small.

**Comment by rkooo567** [comment]
> Would changing the payload size limit not be efficient?

I think limit change is not necessary. But 4MB seems too small anyway (especially given ray default is 512MB). I can revert it if you strongly feel like it.

**Comment by rickyyx** [comment]
> > Would changing the payload size limit not be efficient?
> 
> I think limit change is not necessary. But 4MB seems too small anyway (especially given ray default is 512MB). I can revert it if you strongly feel like it.

I was actually thinking if limit change is enough to tackle the issue?

**Comment by alexeykudinkin** [comment]
@rickyyx limit change by itself is not a solution, right? Irrespective of the limit, workers should not be dropping metrics on the ground -- we have to 

 - Have a control over the batch size (@rkooo567 figured out that we have it, but it turned out to be non-working)
 - Make sure that metrics are reported reliably, even in the presence of failures when we cross the max-size of the batch (which this PR is aiming to address)

**Comment by rickyyx** [comment]
> @rickyyx limit change by itself is not a solution, right? Irrespective of the limit, workers should not be dropping metrics on the ground -- we have to
> 
> * Have a control over the batch size (@rkooo567 figured out that we have it, but it turned out to be non-working)
> * Make sure that metrics are reported reliably, even in the presence of failures when we cross the max-size of the batch (which this PR is aiming to address)

I guess my assumption is that with a larger limit - we will not be dropping metrics already. If that's the case, yes, it's not gonna work with just the limit change.

**Review by rickyyx — APPROVED** [review]
LGTM - leave up to you if you wanna include the nits.

**Comment by rkooo567** [comment]
You can think batch is for the worst of the worst edge case handling. Also if each request has 10+mb, each request will have 100k of metrics which may block thread too long, so break down to multiple requests make sense

**Pull request merged**